cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java

```
@@ -200,4 +245,106 @@ public void getAppDetailForVersion(HttpRequest request, HttpResponder responder,
+          responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, String.format(
+              "Error while reading json request body for app %s.", appId));
+        } catch (Exception e) {
+          LOG.error("Deploy failure", e);
+          responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+        }
+      }
+    };
+  }
+
+  /**
+   * Mark provided application versions as latest
+   *
+   * @param request {@link FullHttpRequest}
+   * @param responder {@link HttpResponse}
+   * @param namespace the namespace where the applications are deployed
+   * @throws Exception if either namespace or any of the application (versions) doesn't exist
+   */
+  @POST
+  @Path("/apps/markLatest")
+  public void markApplicationsAsLatest(FullHttpRequest request, HttpResponder responder,
+      @PathParam("namespace-id") final String namespace) throws Exception {
+    NamespaceId namespaceId = new NamespaceId(namespace);
+    if (!namespaceQueryAdmin.exists(namespaceId)) {
+      throw new NamespaceNotFoundException(namespaceId);
+    }
+
+    AppsMarkLatestRequest appsMarkLatestRequest;
+    try {
+      appsMarkLatestRequest = parseBody(request, AppsMarkLatestRequest.class);
+    } catch (JsonSyntaxException e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Invalid request body: " + e.getMessage());
+      return;
+    }
+
+    List<ApplicationId> appIds = new ArrayList<>();
+    for (AppMarkLatestRequest appRequest: appsMarkLatestRequest.getApps()) {
+      appIds.add(new ApplicationId(namespace, appRequest.getAppId(), appRequest.getVersionId()));
+    }
+
+    try {
+      applicationLifecycleService.markAppsAsLatest(appIds);
+      responder.sendString(HttpResponseStatus.OK, "Success");
+    } catch (MultipleLatestVersionsException e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Failed :" + e.getMessage());
+    } catch (RuntimeException e) {
```

shouldn't catch this, just let it propagate up

---

done